src/Nethermind/Nethermind.Evm/CodeAnalysis/JumpDestinationAnalyzer.cs

```diff
@@ -19,17 +20,56 @@ public sealed class JumpDestinationAnalyzer(ReadOnlyMemory<byte> code)
         private static readonly long[]? _emptyJumpDestinationBitmap = new long[1];
         private long[]? _jumpDestinationBitmap = code.Length == 0 ? _emptyJumpDestinationBitmap : null;
 
+        private object? _analysisComplete;
         private ReadOnlyMemory<byte> MachineCode { get; } = code;
 
         public bool ValidateJump(int destination)
         {
-            ReadOnlySpan<byte> machineCode = MachineCode.Span;
-            _jumpDestinationBitmap ??= CreateJumpDestinationBitmap(machineCode);
+            _jumpDestinationBitmap ??= CreateOrWaitForJumpDestinationBitmap();
 
             // Cast to uint to change negative numbers to very int high numbers
             // Then do length check, this both reduces check by 1 and eliminates the bounds
             // check from accessing the span.
-            return (uint)destination < (uint)machineCode.Length && IsJumpDestination(_jumpDestinationBitmap, destination);
+            return (uint)destination < (uint)MachineCode.Length && IsJumpDestination(_jumpDestinationBitmap, destination);
+        }
+
+        private long[] CreateOrWaitForJumpDestinationBitmap()
+        {
+            object? previous = Volatile.Read(ref _analysisComplete);
+            if (previous is null)
+            {
+                ManualResetEventSlim analysisComplete = new(initialState: false);
+                previous = Interlocked.CompareExchange(ref _analysisComplete, analysisComplete, null);
+                if (previous is null)
+                {
+                    // Not already in progress, so start it.
+                    var bitmap = CreateJumpDestinationBitmap();
+                    _jumpDestinationBitmap = bitmap;
+                    // Release the MRES to be GC'd
+                    _analysisComplete = bitmap;
+                    // Signal complete.
+                    analysisComplete.Set();
+                    return bitmap;
+                }
+            }
+
+            if (previous is ManualResetEventSlim resetEvent)
+            {
+                Thread thread = Thread.CurrentThread;
+                ThreadPriority priority = thread.Priority;
+                // We are waiting, so drop priority to normal (BlockProcessing runs at higher priority).
+                thread.Priority = ThreadPriority.Normal;
+
+                // Already in progress, wait for completion.
+                resetEvent.Wait();
+
+                // Restore the priority of the thread.
+                thread.Priority = priority;
```

Comment:
should this be in finally?

Reply:
This on definately; other not so important as threadpool resets it anyway; however better if we do as that's implementation detail